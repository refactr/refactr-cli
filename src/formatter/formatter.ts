/**
 * Format that is passed to the `--format` option.
 */
export type RawFormatType = 'json' | 'yaml' | 'wide';

export type FormatType = 'json' | 'yaml' | 'table' | 'log';

// TODO(.): shouldn't formatter accept custom formatter per each field?
export type Formatter = (input: unknown, fields: string[]) => string;
