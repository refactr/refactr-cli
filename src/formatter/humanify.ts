import isDate from 'lodash/isDate';
import isBoolean from 'lodash/isBoolean';
import isObject from 'lodash/isObject';
import isNil from 'lodash/isNil';
import parseISO from 'date-fns/parseISO';
import isValid from 'date-fns/isValid';
import formatDate from 'date-fns/format';
import localeEN from 'date-fns/locale/en-US';

const tryParseDate = (value: any): any => {
  const maybeDate = parseISO(value);
  if (isValid(maybeDate)) {
    return maybeDate;
  }

  return value;
};

export default function humanify(value: any) {
  value = tryParseDate(value);

  if (isDate(value)) {
    return humanifyDate(value);
  } else if (isBoolean(value)) {
    return humanifyBoolean(value);
  } else if (isObject(value)) {
    return humanifyObject(value);
  } else if (isNil(value) || value === '') {
    return '-';
  }

  return value;
}

const humanifyDate = (value: Date) =>
  formatDate(value, 'Pp', { locale: localeEN });

const humanifyObject = (value: { [key: string]: any }) => JSON.stringify(value);
const humanifyBoolean = (value: boolean) => (value ? 'Yes' : 'No');
