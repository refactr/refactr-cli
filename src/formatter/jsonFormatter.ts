export default function jsonFormatter(input: any): string {
  try {
    return JSON.stringify(input, null, 4);
  } catch (err) {
    return input;
  }
}
