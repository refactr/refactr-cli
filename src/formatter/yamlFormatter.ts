import { safeDump } from 'js-yaml';

export default function yamlFormatter(input): string {
  return safeDump(input, {
    indent: 4,
  });
}
