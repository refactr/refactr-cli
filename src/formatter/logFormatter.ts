import isNil from 'lodash/isNil';

import humanify from './humanify';
import { Formatter } from './formatter';

const logFormatter: Formatter = function logFormatter(
  input: { [key: string]: unknown },
  fields: string[]
): string {
  return fields
    .map((f) =>
      input[f] === '' || isNil(input[f]) ? null : humanify(input[f])
    )
    .filter((f) => !isNil(f))
    .join(' ');
};

export default logFormatter;
