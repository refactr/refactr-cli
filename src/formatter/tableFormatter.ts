import isArray from 'lodash/isArray';
import isNil from 'lodash/isNil';
import Table from 'cli-table';

import humanify from './humanify';
import { Formatter } from './formatter';

type Row = { [key: string]: any };
type Input = Row | Row[] | null;

const formatterChars = {
  'top': '',
  'top-mid': '',
  'top-left': '',
  'top-right': '',
  'bottom': '',
  'bottom-mid': '',
  'bottom-left': '',
  'bottom-right': '',
  'left': '',
  'left-mid': '',
  'mid': '',
  'mid-mid': '',
  'right': '',
  'right-mid': '',
  'middle': ''
};

const tableFormatter: Formatter = function tableFormatter(
  input: Input,
  fields: string[]
): string {
  if (isNil(input) || input.length === 0) {
    return 'No data returned!';
  }

  if (!isArray(input)) {
    input = [input];
  }

  const rows = (input as Row[]).map((row) =>
    fields.reduce<any[]>((acc, field) => {
      const value = row[field];
      acc.push(humanify(value));

      return acc;
    }, [])
  );

  const table = new Table({
    head: fields,
    chars: formatterChars
  });

  table.push(...rows);

  return table.toString();
};

export default tableFormatter;
