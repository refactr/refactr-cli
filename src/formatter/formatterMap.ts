import type { Formatter, FormatType } from './formatter';
import jsonFormatter from './jsonFormatter';
import yamlFormatter from './yamlFormatter';
import tableFormatter from './tableFormatter';

const formatterMap: Record<FormatType, Formatter> = {
  json: jsonFormatter,
  yaml: yamlFormatter,
  table: tableFormatter,
};

export default formatterMap;
