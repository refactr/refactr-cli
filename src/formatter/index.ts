import isNil from 'lodash/isNil';

import formatterMap from './formatterMap';
import { Formatter, FormatType, RawFormatType } from './formatter';

export type { Formatter, RawFormatType, FormatType };

export const DEFAULT_FORMATTER: RawFormatType = 'wide';

export default function formatter(formatType: FormatType, fields: string[]) {
  return (input: any): string => {
    if (isNil(input)) {
      input = null;
    }

    try {
      return formatterMap[formatType](input, fields);
    } catch (err) {
      return isNil(input) ? 'null' : input!.toString();
    }
  };
}

export function toFormatType(
  format: RawFormatType,
  isLog: boolean = false
): FormatType {
  if (format === 'wide') {
    return isLog ? 'log' : 'table';
  }

  return format;
}
