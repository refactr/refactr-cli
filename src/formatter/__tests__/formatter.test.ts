import formatter from '../index';
import yamlFormatter from '../yamlFormatter';
import jsonFormatter from '../jsonFormatter';
import tableFormatter from '../tableFormatter';

const fields = ['foo', 'bar', 'baz', 'xyz', 'qux', 'quux', 'quuz'];
const data = {
  foo: 'foo',
  bar: 'bar',
  baz: 'baz',
  xyz: null,
  qux: new Date('2021-01-04T13:08:25.618Z'),
  quux: 1.234,
  quuz: {},
};

describe('formatter', () => {
  it('formats to YAML correctly', () => {
    const format = formatter('yaml', fields);

    expect(format(data)).toBe(yamlFormatter(data, fields));
    expect(format(data)).toMatchSnapshot();
  });

  it('formats to JSON correctly', () => {
    const format = formatter('json', fields);

    expect(format(data)).toBe(jsonFormatter(data, fields));
    expect(format(data)).toMatchSnapshot();
  });

  it('formats to table correctly', () => {
    const format = formatter('table', fields);

    expect(format(data)).toBe(tableFormatter(data, fields));
    expect(format(data)).toMatchSnapshot();
  });
});
