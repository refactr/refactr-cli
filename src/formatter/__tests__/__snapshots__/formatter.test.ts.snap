// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatter formats to JSON correctly 1`] = `
"{
    \\"foo\\": \\"foo\\",
    \\"bar\\": \\"bar\\",
    \\"baz\\": \\"baz\\",
    \\"xyz\\": null,
    \\"qux\\": \\"2021-01-04T13:08:25.618Z\\",
    \\"quux\\": 1.234,
    \\"quuz\\": {}
}"
`;

exports[`formatter formats to YAML correctly 1`] = `
"foo: foo
bar: bar
baz: baz
xyz: null
qux: 2021-01-04T13:08:25.618Z
quux: 1.234
quuz: {}
"
`;

exports[`formatter formats to table correctly 1`] = `
"[31m foo [39m[31m bar [39m[31m baz [39m[31m xyz [39m[31m qux                 [39m[31m quux  [39m[31m quuz [39m
 foo  bar  baz  -    01/04/2021, 5:08 AM  1.234  {}   "
`;
